src/crimson/os/seastore/cache.cc

```diff
@@ -1170,12 +1170,18 @@ Cache::get_root_ret Cache::get_root(Transaction &t)
   } else {
     auto ret = root;
     DEBUGT("waiting root {}", t, *ret);
-    return ret->wait_io().then([FNAME, ret, &t] {
+    return ret->wait_io().then([this, FNAME, ret, &t] {
       DEBUGT("got root read {}", t, *ret);
-      t.root = ret;
-      t.add_to_read_set(ret);
-      return get_root_iertr::make_ready_future<RootBlockRef>(
-	ret);
+      if (!ret->is_valid()) {
+	DEBUGT("root became invalid: {}", t, *ret);
+	invalidate(t, *ret);
```

Comment:
I think if the extent is invalid, transactions that are supposed to be invalidated should have been invalidated by the transaction that invalidate the extent, which means we don't have to invalidate them again here, am I right?:-)

Reply:
We don't add root to the read_set until after wait_io(), so no.  Moving the call to add_to_read_set to prior to wait_io would also potentially fix the problem, but wait_io isn't an interruptible future so we'd need to check explicitely anyway.

Reply:
Yes, I agree that we have to check the validy of the root extent after wait_io(), all I'm saying is that we might not have to call `Cache::invalidate()` here, because if an extent is invalid, there must already have been an `Cache::invalidate()` call on that extent, am I right?

Reply:
That particular overload is really only declaring the Transaction conflicted -- it assumes the extent is already invalid.  I'll add a commit renaming the method, this way is indeed confusing.

Reply:
Yes mean this `Cache::invalidate()` call is actually invalidating the current transaction itself, right?

Reply:
Right.  I'm going to rename it to mark_transaction_conflicted.

Reply:
Oh, sorry, I thought it was calling `Cache::invalidate(CachedExtent&)`.

Reply:
Yep, reasonable assumption given the naming.